.github/workflows/push-selfcontained.yml

```
@@ -0,0 +1,90 @@
+name: Push self-contained images on Dockerfile change
+
+on:
+  push:
+    branches:
+    - main
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v1
+    - name: Azure Login
+      uses: Azure/login@v1
+      with:
+        creds: ${{ secrets.AZURE_CREDENTIALS }}
+    - name: Docker Login
+      run: az acr login -n qdkimages
+    - name: Build and push Docker image
+      run: |
+        $Now = [DateTime]::Now;
+        $ImageTag = "${{ github.sha }}"
+
+        $LinuxRepoName = "public/quantum/linux-selfcontained"
+        $WindowsRepoName = "public/quantum/windows-selfcontained"
+
+        $LinuxLocalTag = "${LinuxRepoName}:${ImageTag}-linux"
+        $WindowsLocalTag = "${WindowsRepoName}:${ImageTag}-windows"
+
+        $LinuxRemoteRepo = "${{ secrets.ACR_REGISTRY }}/${LinuxRepoName}"
+        $WindowsRemoteRepo = "${{ secrets.ACR_REGISTRY }}/${WindowsRepoName}"
+
+        docker build images/test-environments/linux --tag $LinuxLocalTag
+        docker build images/test-environments/windows --tag $WindowsLocalTag
+
+        docker tag $LinuxLocalTag "${LinuxRemoteRepo}:${ImageTag}-linux"
+        docker tag $WindowsLocalTag "${WindowsRemoteRepo}:${ImageTag}-windows"
+
+        docker push "${LinuxRemoteRepo}:${ImageTag}-linux"
+        docker push "${WindowsRemoteRepo}:${ImageTag}-windows"
+
+        docker tag $LinuxLocalTag "${LinuxRemoteRepo}:latest"
+        docker tag $WindowsLocalTag "${WindowsRemoteRepo}:latest"
+
+        docker push "${LinuxRemoteRepo}:latest"
+        docker push "${WindowsRemoteRepo}:latest"
```

Nit, but it may be helpful to abstract this into a function to reduce the likelihood of copy-paste errors when this workflow is next modified:
```suggestion
        function Update-Image() {
            param(
                [string] $Path
                [string] $RepoName,
                [string] $LocalTag,
            );
            $fullRepoName = "public/quantum/$RepoName";
            $remoteRepo = "${{ secrets.ACR_REGISTRY }}/${RepoName}";

            docker build $Path --tag $LocalTag;

            docker tag $LocalTag "${remoteRepo}:${LocalTag}";
            docker push "${remoteRepo}:${LocalTag}";

            docker tag $LocalTag "${remoteRepo}:latest";
            docker push "${remoteRepo}:latest";
        }
        Update-Image `
            -RepoName "linux-selfcontained" `
            -LocalTag "${ImageTag}-linux" `
            -Path images/test-environments/linux;
            
        Update-Image `
            -RepoName "windows-selfcontained" `
            -LocalTag "${ImageTag}-windows" `
            -Path images/test-environments/windows;
```

---

Great tip, thank you for refactoring that!